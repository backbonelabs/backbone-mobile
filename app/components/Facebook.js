import React, { Component, PropTypes } from 'react';
import { Alert } from 'react-native';
import {
  AccessToken as FBAccessToken,
  LoginManager,
  GraphRequest,
  GraphRequestManager,
} from 'react-native-fbsdk';
import { connect } from 'react-redux';
import autobind from 'class-autobind';
import constants from '../utils/constants';
import authActions from '../actions/auth';
import styles from '../styles/auth';
import Button from '../components/Button';

class Facebook extends Component {
  static propTypes = {
    buttonText: PropTypes.string,
    dispatch: PropTypes.func,
    navigator: PropTypes.shape({
      push: PropTypes.func,
    }),
  };

  constructor() {
    super();
    autobind(this);
  }

  getFBUserInfo(fbAccessToken) {
    const _responseInfoCallback = (error, result) => {
      if (error) {
        Alert.alert('Please try again.');
      } else {
        // New user object containing request facebook graph fields and login
        // access tokens from facebook
        const user = Object.assign({}, result, fbAccessToken, {
          authMethod: constants.authMethods.FACEBOOK,
        });
        this.props.dispatch(authActions.login(user));
      }
    };

    const infoRequest = new GraphRequest(
      '/me',
      {
        parameters: {
          fields: {
            string: 'email,first_name,last_name,gender,verified',
          },
        },
      },
      _responseInfoCallback
    );

    new GraphRequestManager().addRequest(infoRequest).start();
  }

  render() {
    return (
      <Button
        style={styles._fbBtn}
        textStyle={styles._fbBtnText}
        text={this.props.buttonText}
        fbBtn
        onPress={() =>
          LoginManager.logInWithReadPermissions(['public_profile', 'email']).then(result => {
            if (result && !result.isCancelled) {
              FBAccessToken.getCurrentAccessToken()
                .then((data) => {
                  if (data) {
                    this.getFBUserInfo(data);
                  }
                }
              )
              .catch(() => {
                Alert.alert('Unable to authenticate with Facebook. Try again later.');
              });
            }
          })}
      />
    );
  }
}

const mapStateToProps = state => {
  const { auth } = state;
  return auth;
};

export default connect(mapStateToProps)(Facebook);
