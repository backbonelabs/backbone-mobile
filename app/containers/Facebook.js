import React, { PropTypes } from 'react';
import { Alert } from 'react-native';
import {
  AccessToken as FBAccessToken,
  LoginManager,
  GraphRequest,
  GraphRequestManager,
} from 'react-native-fbsdk';
import { connect } from 'react-redux';
import constants from '../utils/constants';
import authActions from '../actions/auth';
import styles from '../styles/facebook';
import Button from '../components/Button';

/**
 * Collects Facebook user profile data and logs into Backbone API
 * @param {object} fbAccessToken Facebook access token
 */
const getFBUserInfo = (fbAccessToken, dispatch) => {
  const _responseInfoCallback = (error, result) => {
    if (error) {
      Alert.alert('Please try again.');
    } else {
        // Dispatches login with Facebook user profile
      const user = Object.assign({}, result, fbAccessToken, {
        authMethod: constants.authMethods.FACEBOOK,
      });
      dispatch(authActions.login(user));
    }
  };

  const infoRequest = new GraphRequest(
      '/me',
    {
      parameters: {
        fields: {
          string: 'email,first_name,last_name,gender,verified',
        },
      },
    },
      _responseInfoCallback
    );

  new GraphRequestManager().addRequest(infoRequest).start();
};

const login = (dispatch) => {
  LoginManager.logInWithReadPermissions(['public_profile', 'email']).then(result => {
    if (result && !result.isCancelled) {
        // After a Facebook user successfully authenticates, we use the returned Facebook
        // access token to get their profile.
      FBAccessToken.getCurrentAccessToken()
          .then((data) => {
            if (data) {
              getFBUserInfo(data, dispatch);
            }
          }
        )
        .catch(() => {
          Alert.alert('Unable to authenticate with Facebook. Try again later.');
        });
    }
  });
};
const Facebook = (props) => (
  <Button
    style={Object.assign({}, styles._fbBtn, props.style)}
    textStyle={styles._fbBtnText}
    text={props.buttonText}
    fbBtn
    onPress={() => { login(props.dispatch); }}
  />
    );

Facebook.propTypes = {
  dispatch: PropTypes.func,
  buttonText: PropTypes.string,
  style: PropTypes.object,
};

export default { button: connect()(Facebook), login };
