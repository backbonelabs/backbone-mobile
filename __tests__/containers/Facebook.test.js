import React from 'react';
import { shallow } from 'enzyme';
import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import { asyncActionMiddleware } from 'redux-async-action';
import {
  AccessToken as FBAccessToken,
  LoginManager,
} from 'react-native-fbsdk';
import Facebook from '../../app/containers/Facebook';

describe('Facebook Component', () => {
  const store = configureStore([asyncActionMiddleware, thunk])();
  const wrapper = shallow(
    <Facebook buttonText="test" />,
    { context: { store } },
  );
  let render;

  beforeEach(() => {
    store.clearActions();
    render = wrapper.dive();
  });

  test('renders as expected', () => {
    expect(render).toMatchSnapshot();
  });

  test('Login action dispatched when clicked', async () => {
    const btn = render.find('Button').at(0);
    btn.simulate('press');
    await expect(LoginManager.logInWithReadPermissions).toHaveBeenCalled();
    await expect(FBAccessToken.getCurrentAccessToken).toHaveBeenCalled();
  });
});
