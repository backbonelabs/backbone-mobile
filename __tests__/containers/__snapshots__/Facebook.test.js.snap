// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Facebook Component renders as expected 1`] = `
<Button
  fbBtn={true}
  onPress={[Function]}
  shadow={true}
  style={Object {}}
  text="test"
  textStyle={Object {}}
/>
`;
